Source/JavaScriptCore/dfg/DFGObjectAllocationSinkingPhase.cpp

```diff
@@ -2819,6 +2820,47 @@ class ObjectAllocationSinkingPhase : public Phase {
         }
     }
 
+    // The added phis have NodeResultJS because their corresponding Upsilon edges, when coming from nodes in
+    // NamedPropertyPLoc, always have use kinds associated with NodeResultJS. However, this assumption breaks
+    // with the introduction of array allocation sinking, since nodes in ArrayIndexedPropertyPLoc may have
+    // use kinds that produce double results.
+    void fixUpsilonEdge()
+    {
+        for (BasicBlock* block : m_graph.blocksInNaturalOrder()) {
+            for (unsigned indexInBlock = 0; indexInBlock < block->size(); ++indexInBlock) {
+                Node* node = block->at(indexInBlock);
+                switch (node->op()) {
+                case Upsilon: {
+                    Edge& edge = node->child1();
+                    if (node->phi()->result() != resultFor(edge.useKind())) {
+                        ASSERT(node->phi()->hasJSResult());
+
+                        Node* result = nullptr;
+                        if (edge->hasDoubleResult()) {
+                            result = m_insertionSet.insertNode(
+                                indexInBlock, SpecBytecodeDouble, ValueRep, node->origin,
+                                Edge(edge.node(), DoubleRepUse));
+                        } else {
+                            ASSERT(edge->hasInt52Result());
+                            result = m_insertionSet.insertNode(
+                                indexInBlock, SpecInt32Only | SpecAnyIntAsDouble, ValueRep,
+                                node->origin, Edge(edge.node(), Int52RepUse));
+                        }
+
+                        edge.setNode(result);
+                        edge.setUseKind(useKindForResult(NodeResultJS));
+                    }
```

Comment:
To make this more robust, I wonder if we should avoid using `resultFor`.
Like,
```
if (node->phi()->hasJSResult()) {
    Node* result = nullptr;
    if (edge->hasDoubleResult())
        result = m_insertionSet.insertNode(indexInBlock, SpecBytecodeDouble, ValueRep, node->origin, Edge(edge.node(), DoubleRepUse));
    else if (edge->hasInt52Result())
        result = ...

    if (result) {
        edge.setNode(result);
        edge.setUseKind(UntypedUse);
    }
}
```

Reply:
Will change. For these use kinds they must be 
```
inline UseKind useKindForResult(NodeFlags result)
{
    ASSERT(!(result & ~NodeResultMask));
    switch (result) {
    case NodeResultInt52:
        return Int52RepUse;
    case NodeResultDouble:
        return DoubleRepUse;
    default:
        return UntypedUse;
    }
}
```
since we use `defaultEdge()` for inserting Upsilon nodes in `DFGObjectAllocationSinkingPhase.cpp`.
```
m_insertionSet.insertNode(
                        upsilonInsertionPoint, SpecNone, Upsilon, upsilonOrigin,
                        OpInfo(phiNode), incoming->defaultEdge());
```

Reply:
Updated.